src/style-spec/function/type_check.js

```diff
@@ -0,0 +1,290 @@
+// tree with all generics resolved
+function typeCheckExpression(expected: Type, e: TypedExpression) /*: TypedExpression | {| errors: Array<TypeError> |} */ {
+    if (e.literal) {
+        const error = match(expected, e.type);
+        if (error) return { errors: [{ key: e.key, error }] };
+        return e;
+    } else {
+        // e is a lambda expression, so check its result type against the
+        // expected type and recursively typecheck its arguments
+
+        const typenames: { [string]: Type } = {};
+
+        if (expected.kind !== 'lambda') {
+            // if the expected type is not a lambda, then check if it matches
+            // the expression's output type, and then proceed with type checking
+            // the arguments using e.type, which comes from the expression
+            // definition.
+            const error = match(expected, e.type.result, {}, typenames);
+            if (error) return { errors: [{ key: e.key, error }] };
+            expected = e.type;
+        } else {
+            const error = match(expected.result, e.type.result, typenames);
+            if (error) return { errors: [{ key: e.key, error }] };
+        }
+
+        // "Unroll" NArgs if present in the parameter list:
+        // argCount = nargType.type.length * n + nonNargParameterCount
+        // where n is the number of times the NArgs sequence must be
+        // repeated.
+        const argValues = e.arguments;
+        const expandedParams = [];
+        const errors = [];
+        for (const param of expected.params) {
+            if (param.kind === 'nargs') {
+                let count = (argValues.length - (expected.params.length - 1)) / param.types.length;
+                count = Math.min(param.N, Math.ceil(count));
+                while (count-- > 0) {
+                    for (const type of param.types) {
+                        expandedParams.push(type);
+                    }
+                }
+            } else {
+                expandedParams.push(param);
+            }
+        }
+
+        if (expandedParams.length !== argValues.length) {
+            return {
+                errors: [{
+                    key: e.key,
+                    error: `Expected ${expandedParams.length} arguments, but found ${argValues.length} instead.`
+                }]
+            };
+        }
+
+        // Iterate through arguments to:
+        //  - match parameter type vs argument type, checking argument's result type only (don't recursively typecheck subexpressions at this stage)
+        //  - collect typename mappings when ^ succeeds or type errors when it fails
+        for (let i = 0; i < argValues.length; i++) {
+            const param = expandedParams[i];
+            const arg = argValues[i];
+            const error = match(
+                resolveTypenamesIfPossible(param, typenames),
+                arg.type,
+                typenames
+            );
+            if (error) errors.push({ key: arg.key, error });
```

Comment:
Is it valuable to collect & return multiple type checking errors?  If we didn't do this and instead failed at the first error, we could just `throw`, and it would probably make things cleaner.  On the other hand, for expression authors, being able to see & fix multiple errors at once might be nice.